### DISCUSSION TITLE
Callback Error on SignIn with Vercel and Sveltekit

### DISCUSSION THREAD

**Discussion by Logan9312** [nextauthjs/next-auth#8940] [Help]
### Summary

When I do npm run dev locally, everything works perfectly. But when I deploy to vercel and run, I keep getting the following error:

```
[31m[auth][error][CallbackRouteError][0m: Read more at https://errors.authjs.dev#callbackrouteerror
[31m[auth][cause][0m: TypeError: "response" must be an instance of Response
at Module.parseWwwAuthenticateChallenges (file:///var/task/vercel/path0/node_modules/oauth4webapi/build/index.js:586:15)
at handleOAuth (file:///var/task/vercel/path0/node_modules/@auth/core/lib/oauth/callback.js:65:25)
at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
at async Module.callback (file:///var/task/vercel/path0/node_modules/@auth/core/lib/routes/callback.js:20:41)
at async AuthInternal (file:///var/task/vercel/path0/node_modules/@auth/core/lib/index.js:65:38)
at async Auth (file:///var/task/vercel/path0/node_modules/@auth/core/index.js:100:30)
at async respond (file:///var/task/vercel/path0/.svelte-kit/output/server/index.js:2629:22)
at async Server.default (file:///var/task/vercel/path0/.svelte-kit/vercel-tmp/index.js:47:3)
at async Server.<anonymous> (/var/task/___vc/__launcher/__launcher.js:14:11367)
[31m[auth][details][0m: {
"provider": "discord"
}
```

I have been trying to debug this all day, and I don't get what is going wrong. If anyone can help, or at least give me ideas of what I could try that would be greatly appreciated.

Here is my src/hooks.server.ts if that helps

```js
import { SvelteKitAuth } from "@auth/sveltekit";
import Discord from "@auth/core/providers/discord";
import { DISCORD_CLIENT_ID, DISCORD_CLIENT_SECRET } from "$env/static/private";
import type { Handle } from "@sveltejs/kit";

export const handle = SvelteKitAuth({
	providers: [Discord({ clientId: DISCORD_CLIENT_ID, clientSecret: DISCORD_CLIENT_SECRET })],
		callbacks: {
		async jwt({ token, user }) {
			// Destructure the parameters to get `token` and `user`
			if (user) {
				token.id = user.id;
			}
			return token;
		},
		async session({ session, token }: { session: any; token: any }) {
			// Destructure the parameters to get `session` and `token`
			if (session.user) {
				session.user.id = token.id;
				session.user.discriminator = token.discriminator;
			}

			return session;
		},
	}, 
}) satisfies Handle;
```

**Comment by Logan9312** [comment]
After some testing it seems like it’s on Vercel’s end maybe?

I had a branch that was working perfectly fine, and I redeployed and SvelteAuth Sign In stopped working even though I didn’t change anything. Any ideas how I could fix this?

**Comment by johnyvelho** [comment]
same here. I am also debugging but no success so far. It works fine locally (also preview), but it doesn't on vercel.
vercel + sveltekit + google as provider:

error: 

`[31m[auth][error][SignInError][0m: Read more at https://errors.authjs.dev#signinerror
[31m[auth][cause][0m: TypeError: "response" must be an instance of Response
at Module.processDiscoveryResponse (file:///var/task/vercel/path0/node_modules/oauth4webapi/build/index.js:212:15)
at getAuthorizationUrl (file:///var/task/vercel/path0/node_modules/@auth/core/lib/oauth/authorization-url.js:19:28)
at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
at async Module.signin (file:///var/task/vercel/path0/node_modules/@auth/core/lib/routes/signin.js:15:20)
at async AuthInternal (file:///var/task/vercel/path0/node_modules/@auth/core/lib/index.js:112:36)
at async Auth (file:///var/task/vercel/path0/node_modules/@auth/core/index.js:100:30)
at async respond (file:///var/task/vercel/path0/.svelte-kit/output/server/index.js:2556:22)
at async Server.default (file:///var/task/vercel/path0/.svelte-kit/vercel-tmp/index.js:47:3)
at async Server.<anonymous> (/var/task/___vc/__launcher/__launcher.js:14:11367)
[31m[auth][details][0m: {
"provider": "google"
}`

My error is about "signinerror", while your error is about "callbackrouteerror". I am using google as provider

**Comment by johnyvelho** [comment]
after replicating the @auth/core in my project,  pointing to it, and console.logging in order to be able to get the logs on vercel enviorment, I was able to find out that the errors occours during the invocation of the method processDiscoveryResponse from the package "oauth4webapi".

file: @auth/core/lib/oauth/authorization-url.js
function: getAuthorizationUrl
when calling: `const as = await o.processDiscoveryResponse(issuer, discoveryResponse);`

I compared the vars "issuer" and "discoveryResponse" from local and vercel output. both are exactly the same.
as we can see in the logs below, it didn't hit the "ENTERED:1.4"

<img width="733" alt="image" src="https://github.com/nextauthjs/next-auth/assets/5813213/6e23110a-b907-4a32-bf66-6c30c6e86829">

`ENTERED
ENTERED:1:1
ENTERED:1:2
ENTERED:1:3 URL {
href: 'https://accounts.google.com/',
origin: 'https://accounts.google.com',
protocol: 'https:',
username: '',
password: '',
host: 'accounts.google.com',
hostname: 'accounts.google.com',
port: '',
pathname: '/',
search: '',
searchParams: URLSearchParams {},
hash: ''
} Response {
[Symbol(realm)]: null,
[Symbol(state)]: {
aborted: false,
rangeRequested: false,
timingAllowPassed: true,
requestIncludesCredentials: true,
type: 'default',
status: 200,
timingInfo: {
startTime: 20070.035333000007,
redirectStartTime: 0,
redirectEndTime: 0,
postRedirectStartTime: 20070.035333000007,
finalServiceWorkerStartTime: 0,
finalNetworkResponseStartTime: 0,
finalNetworkRequestStartTime: 0,
endTime: 0,
encodedBodySize: 464,
decodedBodySize: 0,
finalConnectionTimingInfo: null
},
cacheState: '',
statusText: 'OK',
headersList: HeadersList {
cookies: null,
[Symbol(headers map)]: [Map],
[Symbol(headers map sorted)]: null
},
urlList: [ [URL] ],
body: { stream: undefined }
},
[Symbol(headers)]: HeadersList {
cookies: null,
[Symbol(headers map)]: Map(18) {
'accept-ranges' => [Object],
'content-encoding' => [Object],
'access-control-allow-origin' => [Object],
'content-security-policy-report-only' => [Object],
'cross-origin-opener-policy' => [Object],
'report-to' => [Object],
'content-length' => [Object],
'x-content-type-options' => [Object],
'server' => [Object],
'x-xss-protection' => [Object],
'date' => [Object],
'expires' => [Object],
'cache-control' => [Object],
'age' => [Object],
'last-modified' => [Object],
'content-type' => [Object],
'vary' => [Object],
'alt-svc' => [Object]
},
[Symbol(headers map sorted)]: null
}
}
[31m[auth][error][SignInError][0m: Read more at https://errors.authjs.dev#signinerror
[31m[auth][cause][0m: TypeError: "response" must be an instance of Response
at Module.processDiscoveryResponse (file:///var/task/vercel/path0/node_modules/oauth4webapi/build/index.js:212:15)
at getAuthorizationUrl (file:///var/task/vercel/path0/node_modules/@auth/core/lib/oauth/authorization-url.js:24:22)
at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
at async Module.signin (file:///var/task/vercel/path0/node_modules/@auth/core/lib/routes/signin.js:16:17)
at async AuthInternal (file:///var/task/vercel/path0/node_modules/@auth/core/lib/index.js:112:36)
at async Auth (file:///var/task/vercel/path0/node_modules/@auth/core/index.js:101:30)
at async respond (file:///var/task/vercel/path0/.svelte-kit/output/server/index.js:2556:22)
at async Server.default (file:///var/task/vercel/path0/.svelte-kit/vercel-tmp/index.js:47:3)
at async Server.<anonymous> (/var/task/___vc/__launcher/__launcher.js:14:11367)
[31m[auth][details][0m: {
"provider": "google"
}`

my investigation goes until here. I will continue later

**Comment by Logan9312** [comment]
Okay so i tried using the following versions of SvelteAuth:     "@auth/core": "0.10.0", "@auth/sveltekit": "0.3.6",

Doing this makes my auth work properly again, so until a fix is found, these versions seem good enough to work

**Reply by trondhauklien** [reply]
Had the same issue. Downgrading seems to fix it.

**Reply by sneakyPad** [reply]
Thank you! That worked for me, although I feel a bit uncomfortable about being stuck with those dependencies. There's another thread discussing downgrading Nodejs from 18 to 16, but I haven't tested it. [next-auth/issues/8943](https://github.com/nextauthjs/next-auth/issues/8943#issue-1960825828)

**Comment by pupvogel** [comment]
Hm, this doesn't seem to work for me - same problem as before when I change my package.json to use these versions in vercel..!?

**Reply by Logan9312** [reply]
Is it possible that you didn't update your lockfile after changing the version dependencies in package.json? Maybe for some reason Vercel is still trying to use the versions you had installed before changing it

**Reply by pupvogel** [reply]
Sorry, no, the problem just disappeared later that day, looks like it was caused by something on the Vercel-end.

Same stuff that didn't work before works now. Sometimes we just have to wait I guess... 8o)

**Comment by johnyvelho** [comment]
I hit a dead end.
this is the function causing the issue from oauth4webapi:

```javascript
export async function processDiscoveryResponse(expectedIssuerIdentifier, response) {
    if (!(expectedIssuerIdentifier instanceof URL)) {
        throw new TypeError('"expectedIssuer" must be an instance of URL');
    }
    if (!(response instanceof Response)) {
        throw new TypeError('"response" must be an instance of Response');
    }
    if (response.status !== 200) {
        throw new OPE('"response" is not a conform Authorization Server Metadata response');
    }
    assertReadableResponse(response);
    let json;
    try {
        json = await response.json();
    }
    catch {
        throw new OPE('failed to parse "response" body as JSON');
    }
    if (!isJsonObject(json)) {
        throw new OPE('"response" body must be a top level object');
    }
    if (!validateString(json.issuer)) {
        throw new OPE('"response" body "issuer" property must be a non-empty string');
    }
    if (new URL(json.issuer).href !== expectedIssuerIdentifier.href) {
        throw new OPE('"response" body "issuer" does not match "expectedIssuer"');
    }
    return json;
}
```

the error is coming from this block:
`if (!(response instanceof Response)) {
        throw new TypeError('"response" must be an instance of Response');
    }
`

printing the constructor name both in local and Vercel shows 'Response'. however, in production this condition returns false for some reason. I went back several versions on `@auth/core` and `oauth4webapi` and at least this part seems the same.
Since locally works normally, I am assuming Vercel is changing the Response somehow in order to change the instanceof Response.
I am considering change to another host or Lucia js.

the mainterners probably would have a good idea what the issue is, so maybe they will help us.

**Reply by sneakyPad** [reply]
Thanks for the explanation! Since this happened to me on prod, I moved my application to Netlify. It was the quickest solution but I like Vercel more so I'll try to go back there unless it happens again. Then I will have to move to another host.
